File: JUNE-2025/28-06-2025/p2099.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer[]}
def max_subsequence(nums, k)
    sorted_nums = nums.sort

    hash = Hash.new(0)
    sorted_nums[nums.size - k..-1].each do |num|
        hash[num] += 1
    end

    answer = []
    nums.each do |num|
        if hash[num] > 0
            hash[num] -= 1
            answer << num
        end
    end

    answer
end

File: JUNE-2025/28-06-2025/p2099.php
class Solution {

    /**
     * @param Integer[] $nums
     * @param Integer $k
     * @return Integer[]
     */
    function maxSubsequence($nums, $k) {
        if( $k === count( $nums ) ) return $nums;
        
        $copy = $nums; // Make a copy not to mod original arr
        rsort( $copy ); // Reverse sort for get the k largest
        
        // Build the frequency mamp
        $freq = array_count_values( array_splice( $copy, 0, $k ) );
        
        $res = [];
        foreach( $nums as $num ) {
            if( isset( $freq[$num] ) && $freq[$num] > 0 ) {
                $res[] = $num;
                $freq[$num]--;
                if (count($res) === $k) break;
            }
        }
        
        return $res;
    }
}